source/extensions/filters/http/set_metadata/set_metadata_filter.cc

```diff
@@ -13,23 +13,82 @@ namespace Extensions {
+               Stats::Scope& scope, const std::string& stats_prefix)
+    : stats_(generateStats(stats_prefix, scope)) {
+  if (proto_config.has_value() && !proto_config.metadata_namespace().empty()) {
+    UntypedMetadataEntry deprecated_api_val{true, proto_config.metadata_namespace(),
+                                            proto_config.value()};
+    untyped_.emplace_back(deprecated_api_val);
+  }
+
+  for (const auto& metadata : proto_config.metadata()) {
+    if (metadata.has_value()) {
+      UntypedMetadataEntry untyped_entry{metadata.allow_overwrite(), metadata.metadata_namespace(),
+                                         metadata.value()};
+      untyped_.emplace_back(untyped_entry);
+    } else if (metadata.has_typed_value()) {
+      TypedMetadataEntry typed_entry{metadata.allow_overwrite(), metadata.metadata_namespace(),
+                                     metadata.typed_value()};
+      typed_.emplace_back(typed_entry);
+    } else {
+      ENVOY_LOG(warn, "set_metadata filter configuration contains metadata entries without value "
+                      "or typed_value");
+    }
+  }
+}
+
+FilterStats Config::generateStats(const std::string& prefix, Stats::Scope& scope) {
+  std::string final_prefix = prefix + "set_metadata.";
+  return {ALL_SET_METADATA_FILTER_STATS(POOL_COUNTER_PREFIX(scope, final_prefix))};
 }
 
 SetMetadataFilter::SetMetadataFilter(const ConfigSharedPtr config) : config_(config) {}
 
 SetMetadataFilter::~SetMetadataFilter() = default;
 
 Http::FilterHeadersStatus SetMetadataFilter::decodeHeaders(Http::RequestHeaderMap&, bool) {
-  const absl::string_view metadata_namespace = config_->metadataNamespace();
-  auto& metadata = *decoder_callbacks_->streamInfo().dynamicMetadata().mutable_filter_metadata();
-  ProtobufWkt::Struct& org_fields =
-      metadata[toStdStringView(metadata_namespace)]; // NOLINT(std::string_view)
-  const ProtobufWkt::Struct& to_merge = config_->value();
 
-  StructUtil::update(org_fields, to_merge);
+  // add configured untyped metadata
+  if (!config_->untyped().empty()) {
+    auto& mut_untyped_metadata =
+        *decoder_callbacks_->streamInfo().dynamicMetadata().mutable_filter_metadata();
+
+    for (const auto& entry : config_->untyped()) {
+      if (!mut_untyped_metadata.contains(entry.metadata_namespace)) {
+        mut_untyped_metadata[entry.metadata_namespace] = entry.value;
+      } else if (entry.allow_overwrite) {
+        // get the existing metadata at this key for merging
+        ProtobufWkt::Struct& orig_fields = mut_untyped_metadata[entry.metadata_namespace];
+        const auto& to_merge = entry.value;
+
+        // merge the new metadata into the existing metadata
+        StructUtil::update(orig_fields, to_merge);
+      } else {
+        // entry exists, not allowed to overwrite -- emit a stat
+        config_->stats().overwrite_denied_.inc();
+      }
+    }
+  }
```

Comment:
Could you revise the format of these code comments? First letter should be uppercase, end with dot, grammar, etc.